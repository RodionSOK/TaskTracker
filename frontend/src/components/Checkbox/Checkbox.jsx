import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Icon from "../Icon/Icon.jsx";

import "./Checkbox.pcss";

const Checkbox = ({
    className,
    id,
    value,
    checked,
    defaultChecked,
    label,
    labelCount,
    tabIndex,
    theme,
    error,
    disabled,
    onChange,
    onClick,
    onBlur,
}) => {
    const componentClassName = classNames("checkbox", className, {
        checkbox_white: theme === "white",
        "checkbox_white-with-border": theme === "white-with-border",
        checkbox_error: error,
    });

    return (
        <div onClick={onClick} className={componentClassName}>
            <input
                className="checkbox__input"
                id={id}
                type="checkbox"
                disabled={disabled}
                tabIndex={tabIndex}
                onChange={onChange}
                onBlur={onBlur}
                value={value}
                checked={checked}
                defaultChecked={defaultChecked}
            />
            <label htmlFor={id} className="checkbox__label">
                <Icon className="checkbox__icon" type="mail-check" width="14" height="14" />
                {label}

                {labelCount && <span className="checkbox__label-count">{labelCount}</span>}
            </label>
        </div>
    );
};

Checkbox.propTypes = {
    /** Id */
    id: PropTypes.string,

    /** Дополнительный класс */
    className: PropTypes.string,

    /** Value */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),

    /** Checked */
    checked: PropTypes.bool,

    /** Значение по умолчанию */
    defaultChecked: PropTypes.bool,

    /** Текст */
    label: PropTypes.string,

    /** Дополнительный текст (например, вывод количества результатов с таким фильтром) */
    labelCount: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

    /** Тема */
    theme: PropTypes.oneOf(["white", "white-with-border"]),

    /** Tab index */
    tabIndex: PropTypes.string,

    /** Вид с ошибкой */
    error: PropTypes.bool,

    /** Заблокированное состояние */
    disabled: PropTypes.bool,

    /** Обработчик выбора */
    onChange: PropTypes.func,

    /** Обработчик нажатия */
    onClick: PropTypes.func,

    onBlur: PropTypes.func,
};

export default Checkbox;
