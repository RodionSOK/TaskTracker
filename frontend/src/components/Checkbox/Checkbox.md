Ch1. Чекбокс. Белый

```jsx
initialState = { value: false };

<div className="styleguide__labeled styleguide_bg styleguide_bg-gray">
    <Checkbox
        id="white-checkbox-default"
        theme="white"
        label="Белый чекбокс"
        tabIndex="1"
        value={state.value}
        onChange={() => setState({ value: !state.value })}
    />
</div>;
```

Ch2. Чекбокс. Серый

```jsx
initialState = { value: false };

<div className="styleguide__labeled">
    <Checkbox
        id="basic-checkbox-default"
        label="Серый чекбокс"
        tabIndex="2"
        value={state.value}
        onChange={() => setState({ value: !state.value })}
    />
</div>;
```

Ch3. Чекбокс. Белый с бордером

```jsx
initialState = { value: false };

<div className="styleguide__labeled">
    <Checkbox
        id="white-with-border-сeckbox-default"
        theme="white-with-border"
        label="Белый чекбокс с бордером"
        tabIndex="3"
        value={state.value}
        onChange={() => setState({ value: !state.value })}
    />
</div>;
```

Ch3. Чекбокс. С дополнительным текстом

```jsx
initialState = { value: false };

<div className="styleguide__labeled">
    <Checkbox
        id="white-with-border-сeckbox-count"
        theme="white-with-border"
        label="Белый чекбокс с дополнитеьный текстом"
        labelCount="12"
        tabIndex="4"
        value={state.value}
        onChange={() => setState({ value: !state.value })}
    />
</div>;
```
