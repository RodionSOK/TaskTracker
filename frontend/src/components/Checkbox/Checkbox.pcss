@import "/src/styles/settings.pcss";

.checkbox {
    font-size: 14px;
    line-height: 18px;
    padding-left: 16px;

    &_white {
        .checkbox__label {
            &::before {
                background-color: var(--white);
            }
        }

        .checkbox__label:hover {
            &::before {
                background-color: var(--gray_90);
            }
        }

        .checkbox__input:disabled + .checkbox__label:hover::before {
            background-color: var(--white);
        }
    }

    &_white-with-border {
        padding-left: 22px;
        min-height: 22px;

        .checkbox__input {
            &:checked + .checkbox__label {
                &::before {
                    border-color: var(--black);
                    background-color: var(--black);
                }
            }

            &:disabled + .checkbox__label {
                color: var(--gray_50);
            }
            &:disabled + .checkbox__label:hover::before {
                background-color: var(--white);
                border-color: var(--gray_90);
            }

            &:focus + .checkbox__label::before {
                box-shadow: none;
                border-color: var(--black);
            }

            &:disabled:checked + .checkbox__label:hover::before {
                border-color: var(--black);
                background-color: var(--black);
            }
        }

        .checkbox__label {
            padding-top: 3px;

            &:hover {
                &::before {
                    border-color: var(--gray_80);
                    background-color: var(--white);
                }
            }

            &::before {
                top: 1px;
                left: -21px;
                width: 20px;
                height: 20px;
                border: 1px solid var(--gray_90);
                background-color: var(--white);
            }
        }

        .checkbox__icon {
            top: 5px;
            left: -17px;
            fill: var(--white);
        }
    }

    &.checkbox_white-with-border.checkbox_error {
        .checkbox__label {
            &::before {
                border-color: var(--error);
                box-shadow: none;
            }
        }
    }

    &_error {
        .checkbox__label {
            &::before {
                box-shadow: inset 0 0 0 1px var(--error);
            }
        }
    }

    &__input {
        position: absolute;
        left: -1000em;
        opacity: 0;
        width: 0;
        outline: none;
        display: none;

        &:checked + .checkbox__label {
            &::before {
                background-color: var(--green);
            }

            .checkbox__icon {
                display: block;
            }
        }

        &:disabled + .checkbox__label {
            cursor: not-allowed;
        }

        &:disabled:checked + .checkbox__label {
            cursor: pointer;
        }

        &:disabled + .checkbox__label:hover::before {
            background-color: var(--gray_90);
        }

        &:focus + .checkbox__label::before {
            box-shadow: inset 0 0 0 1px var(--green);
        }
    }

    &__label {
        display: inline-block;
        position: relative;
        font-size: inherit;
        font-family: var(--base-font-family);
        font-weight: var(--base-font-weight);
        outline: none;
        margin: 0;
        padding-left: 10px;
        cursor: pointer;
        user-select: none;

        &:hover {
            &::before {
                background-color: var(--gray_85);
            }
        }

        &::before {
            content: "";
            position: absolute;
            left: -15px;
            width: 16px;
            height: 16px;
            top: 1px;
            background-color: var(--gray_90);
            transition: var(--tr_time);
        }
    }

    &__label-count {
        color: var(--gray_50);
        margin-left: 5px;
    }

    &__icon {
        display: none;
        position: absolute;
        top: 1px;
        left: -14px;
        fill: var(--white);
    }
}

.checkbox-list {
    .checkbox {
        margin-bottom: 8px;
    }
}
