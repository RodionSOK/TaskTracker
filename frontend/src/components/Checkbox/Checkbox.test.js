import React from "react";
import Checkbox from "./Checkbox.jsx";
import renderer from "react-test-renderer";

it("default checkbox", () => {
    const tree = renderer.create(<Checkbox id="basic-checkbox-default" label="Серый чекбокс" tabIndex="1" />).toJSON();
    expect(tree).toMatchSnapshot();
});

it("checkbox with white theme", () => {
    const tree = renderer
        .create(<Checkbox id="white-checkbox-default" theme="white" label="Белый чекбокс" tabIndex="1" />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("checkbox with error", () => {
    const tree = renderer
        .create(<Checkbox id="basic-checkbox-default" label="Серый чекбокс" tabIndex="1" error />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
