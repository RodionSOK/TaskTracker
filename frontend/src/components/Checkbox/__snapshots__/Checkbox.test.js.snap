// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox with error 1`] = `
<div
  className="checkbox checkbox_error"
  onClick={undefined}
>
  <input
    checked={undefined}
    className="checkbox__input"
    defaultChecked={undefined}
    disabled={undefined}
    id="basic-checkbox-default"
    onBlur={undefined}
    onChange={undefined}
    tabIndex="1"
    type="checkbox"
    value={undefined}
  />
  <label
    className="checkbox__label"
    htmlFor="basic-checkbox-default"
  >
    <svg
      className="icon checkbox__icon"
      height="14"
      id={undefined}
      onClick={undefined}
      width="14"
    >
      <use
        xlinkHref="#mail-check"
      />
    </svg>
    Серый чекбокс
  </label>
</div>
`;

exports[`checkbox with white theme 1`] = `
<div
  className="checkbox checkbox_white"
  onClick={undefined}
>
  <input
    checked={undefined}
    className="checkbox__input"
    defaultChecked={undefined}
    disabled={undefined}
    id="white-checkbox-default"
    onBlur={undefined}
    onChange={undefined}
    tabIndex="1"
    type="checkbox"
    value={undefined}
  />
  <label
    className="checkbox__label"
    htmlFor="white-checkbox-default"
  >
    <svg
      className="icon checkbox__icon"
      height="14"
      id={undefined}
      onClick={undefined}
      width="14"
    >
      <use
        xlinkHref="#mail-check"
      />
    </svg>
    Белый чекбокс
  </label>
</div>
`;

exports[`default checkbox 1`] = `
<div
  className="checkbox"
  onClick={undefined}
>
  <input
    checked={undefined}
    className="checkbox__input"
    defaultChecked={undefined}
    disabled={undefined}
    id="basic-checkbox-default"
    onBlur={undefined}
    onChange={undefined}
    tabIndex="1"
    type="checkbox"
    value={undefined}
  />
  <label
    className="checkbox__label"
    htmlFor="basic-checkbox-default"
  >
    <svg
      className="icon checkbox__icon"
      height="14"
      id={undefined}
      onClick={undefined}
      width="14"
    >
      <use
        xlinkHref="#mail-check"
      />
    </svg>
    Серый чекбокс
  </label>
</div>
`;
